From 143d6a42ac86ebbf51cb4f440519d42bb38c55fd Mon Sep 17 00:00:00 2001
From: Wendy Liang <wendy.liang@xilinx.com>
Date: Fri, 4 Sep 2020 16:17:04 -0700
Subject: [PATCH 1568/1852] misc: xilinx-ai-engine: fix scanning gated tiles
 logic

commit b074fd34c174890e785891cf4c2d156c806718f4 from
https://github.com/Xilinx/linux-xlnx.git

Fix the logic to scan the gated tiles.
* remove additional setting va to check shim clock gate status.
* if the next tile is gated, skip the rest rows in the column.

Signed-off-by: Wendy Liang <wendy.liang@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/misc/xilinx-ai-engine/ai-engine-aie.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-aie.c b/drivers/misc/xilinx-ai-engine/ai-engine-aie.c
index ed45d7a59bee..ab4f6956be7e 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-aie.c
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-aie.c
@@ -272,7 +272,6 @@ static int aie_scan_part_clocks(struct aie_partition *apart)
 				va = adev->base +
 				     aie_cal_regoff(adev, loc,
 						    AIE_SHIMPL_CLKCNTR_REGOFF);
-				va = adev->base + AIE_SHIMPL_CLKCNTR_REGOFF;
 				val = ioread32(va);
 
 				/*
@@ -296,10 +295,14 @@ static int aie_scan_part_clocks(struct aie_partition *apart)
 					    AIE_TILE_CORE_CLKCNTR_REGOFF);
 			val = ioread32(va);
 
-			if (val & AIE_TILE_CLKCNTR_NEXTCLK_MASK) {
-				aie_resource_set(&apart->cores_clk_state,
-						 nbitpos, 1);
-			}
+			/*
+			 * If the next tile is gated, skip the rest of the
+			 * column.
+			 */
+			if (!(val & AIE_TILE_CLKCNTR_NEXTCLK_MASK))
+				break;
+
+			aie_resource_set(&apart->cores_clk_state, nbitpos, 1);
 		}
 	}
 
-- 
2.31.1

